password.c

```diff
@@ -0,0 +1,227 @@
+#define malloc  mock_malloc
+#define free    mock_free
+#else
+#endif
+
+/* Macros for have_next_line result */
+#define HAVE_NEXT_SETTING_LINE 1
+#define NO_NEXT_SETTING_LINE   0
+
+/* Macros for parse user input */
+#define USER_COMMAND_TOKEN_WHITESPACE              " \t\n\r\f"
+#define USER_COMMAND_TOKEN_SETTING_SPLITTER        " =\t"
+#define USER_COMMAND_TOKEN_EQUAL                   "="
+#define USER_COMMAND_TOKEN_COMMA                   ","
+
+/* The command alias prefix */
+const char* COMMAND_ALIAS = "Cmnd_Alias";
+
+/* The password setting */
+const char* PASSWORD_SETTING = "PASSWD_CMDS";
+
+/* Regex list */
+REGEX_NODE *global_regex_list = NULL;
+
+/* Append regex to list */
+int append_regex_to_list(regex_t regex)
+{
+    /* Create and initialize regex node */
+    REGEX_NODE *new_regex_node = (REGEX_NODE *)malloc(sizeof(REGEX_NODE));
+    if (new_regex_node == NULL)
+    {
+        /* When allocate memory failed, stop and return. also output log to both syslog and stderr with LOG_PERROR*/
+        trace("Failed to allocate memory for regex node.\n");
+        return REGEX_APPEND_FAILED;
+    }
+
+    new_regex_node->next = NULL;
+    new_regex_node->regex = regex;
+
+    /* Find the pointer to the latest regex node's 'next' field */
+    REGEX_NODE **current_node = &global_regex_list;
+    while (*current_node != NULL) {
+        current_node = &((*current_node)->next);
+    }
+    
+    /* append new regex to tail node */
+    *current_node = new_regex_node;
+    return REGEX_APPEND_SUCCESS;
+}
+
+/* Release password setting */
+void release_password_setting()
+{
+    if (global_regex_list == NULL) {
+        return;
+    }
+
+    /* Walk to last regex */
+    REGEX_NODE *current = global_regex_list;
+    while (current != NULL) {
+        /* Continue with next regex */
+        REGEX_NODE* current_node_memory = current;
+        current = current->next;
+        
+        /* Free node memory, this may also reset all allocated memory depends on c lib implementation */
+        free(current_node_memory);
+    }
+
+    /* Reset list */
+    global_regex_list = NULL;
+}
+
+/* Replace password with regex */
+void remove_password(char* command)
+{
+    if (global_regex_list == NULL) {
+        return;
+    }
+
+    /* Check every regex */
+    REGEX_NODE *next_node = global_regex_list;
+    while (next_node != NULL) {
+        /* Try fix password with current regex */
+        if (remove_password_by_regex(command, next_node->regex) == PASSWORD_REMOVED) {
+            return;
+        }
+        
+        /* If password not fix, continue try next regex */
+        next_node = next_node->next;
+    }
+}
+
+/* Find and return the pointer of the first none space character*/
+char* find_non_space(char *str)
+{
+    if (str == NULL) {
+        return str;
+    }
+
+    while (isspace(*str)) {
+        str++;
+    }
+
+    return str;
+}
+
+/* Check setting if have next line */
+int check_have_next_line(const char *str)

```

Comment:
>check_have_next_line [](http://example.com/codeflow?start=4&length=20)

Add some unit test?

Reply:
Fixed.
This method been removed, because I realize the old code can't handle sudoers file format correctly.
Now I created a new file sudoers_helper.c and move all sudoers file related code there, include escape characters handling and parse PASSWD_CMDS.

Reply:
Fixed.
This method been removed, because I realize the old code can't handle sudoers file format correctly.
Now I created a new file sudoers_helper.c and move all sudoers file related code there, include escape characters handling and parse PASSWD_CMDS.